backend/drm/backend.c

```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
 	return b->impl == &backend_impl;
 }
 
+static void wlr_drm_connector_update_crtc_binding(struct wlr_drm_connector *conn,
+		struct wlr_drm_backend *drm) {
+	drmModeConnectorPtr drm_conn = drmModeGetConnector(drm->fd, conn->id);
+	if (!drm_conn) {
+		wlr_log_errno(L_ERROR, "Failed to get DRM connector %u", conn->id);
+		return;
+	}
+	if (!drm_conn->encoder_id) {
+		conn->crtc = NULL;
+		drmModeFreeCrtc(conn->old_crtc);
+		conn->old_crtc = NULL;
+		conn->state = WLR_DRM_CONN_DISCONNECTED;
+
+		drmModeFreeConnector(drm_conn);
+		return;
+	}
+
+	drmModeEncoderPtr drm_encoder = drmModeGetEncoder(drm->fd,
+		drm_conn->encoder_id);
+	if (!drm_encoder) {
+		wlr_log_errno(L_ERROR, "Failed to get DRM encoder %u on connector %u",
+			drm_conn->encoder_id, conn->id);
+
+		drmModeFreeConnector(drm_conn);
+		return;
+	}
+
+	if (!conn->crtc || drm_encoder->crtc_id != conn->crtc->id) {
+		drmModeFreeCrtc(conn->old_crtc);
+		conn->old_crtc = drmModeGetCrtc(drm->fd, drm_encoder->crtc_id);
+
+		for (size_t i = 0; i < drm->num_crtcs; ++i) {
+			if (drm->crtcs[i].id == drm_encoder->crtc_id) {
+				conn->crtc = &drm->crtcs[i];
+			}
+		}
+	}
+
+	drmModeFreeEncoder(drm_encoder);
+	drmModeFreeConnector(drm_conn);
+}
+
+static void wlr_drm_backend_update_connector_crtc_bindings(
```

Comment:
I wouldn't bother with a function this small, especially if it's only going to be used in one place.
I would just write the for loop where it's going to be used.